docs/testing/integration-tests.md

```diff
@@ -0,0 +1,219 @@
+
+## Unit Tests vs. Integration Tests
+
+Unit tests are great for ensuring that functionality works as expected at an insolated, functional-level. This might include checking that function `get_nth_fibonacci(n: u8)` works as expected, handles invalid input gracefully, etc. Unit tests in smart contracts might similarly test public functions, but can get unruly if there are several calls between accounts. As mentioned in the [unit tests](unit-tests.md) section, there is a `VMContext` object used by unit tests to mock some aspects of a transaction. One might, for instance, modify the testing context to have the `predecessor_account_id` of `"bob.near"`. The limits of unit tests become obvious with certain interactions, like transferring tokens. Since `"bob.near"` is simply a string and not an account object, there is no way to write a unit test that confirms that Alice sent Bob 6 NEAR (Ⓝ). Furthermore, there is no way to write a unit test that executes cross-contract calls. Additionally, there is no way of profiling gas usage and the execution of the call (or set of calls) on the blockchain.
+
+Integration tests provide the ability to have end-to-end testing that includes cross-contract calls, proper user accounts, access to state, structured execution outcomes, and more. In NEAR, we can make use of the `workspaces` libraries in both [Rust](https://github.com/near/workspaces-rs) and [JavaScript](https://github.com/near/workspaces-js) for this type of testing on a locally-run blockchain or testnet.
+
+## When to Use Integration Tests
+
+You'll probably want to use integration tests when:
+
+- There are cross-contract calls.
+- There are multiple users with balance changes.
+- You'd like to gather information about gas usage and execution outcomes on-chain.
+- You want to assert the use-case execution flow of your smart contract logic works as expected.
+- You want to assert given execution patterns do not work (as expected).
+
+## Setup
+
+Unlike unit tests (which would often live in the `src/lib.rs` file of the contract), integration tests in Rust are located in a separate directory at the same level as `/src`, called `/tests` ([read more](https://doc.rust-lang.org/cargo/reference/cargo-targets.html#integration-tests)). Refer to this folder structure below:
```

Comment:
They don't have to be. Can have workspaces tests within the src directory too. This can be beneficial because then you can use the non-exported types for serialization within the test case.

Reply:
I added an `info` clause below mentioning this, thank you!